src/main/golo/functions.golo

```diff
@@ -0,0 +1,871 @@
+----
+This module contains several functions and higher order functions or
+decorators, mostly useful with HOF like `map`, `reduce` or in generators.
+
+Indeed, using HOF often require some classical functions like constants one,
+identity, or classical operators.
+
+Most of the times, these functions needs to be unary ones, i.e.
+have only one parameter. This module contains, among others, functions helping in the arity
+conversion of such functions:
+
+* [`curry`](#curry_f) that makes a function automatically partializable,
+* [`uncurry`](#uncurry_f) that is the reverse of `curry`,
+* [`unary`](#unary_f) and [`spreader`](#spreader_f) that convert a polyadic function into a unary
+  function taking an array or a tuple of values instead of several
+  values,
+* [`varargs`](#varargs_f) which is the reverse of `unary`
+
+Most of the binary operator functions defined here are curried. To ease
+partialization, the non commutative operators have a reversed version, with
+arguments swapped.
+For example, using the reverse division operator [`rdiv`](#rdiv_x_y) with
+[`map`](StandardAugmentations.html#java.util.List.map_this_func)
+
+    [2, 4, 8]: map(rdiv(2)) == [1, 2, 4]
+
+The identity function [`id`](#id_x), the constant function
+[`const`](#const_val), together with composition (`FunctionReference:andThen`,
+[`pipe`](#pipe_funcs), [`compose`](#compose_funcs)) and [`io`](#io_block) can be
+used to create monad-like processing chains.
+
+----
+module gololang.functions
+
+import java.lang.invoke
+import java.util
+
+### Operators ##########################################################
+
+#== Arithmetic =========================================================
+----
+Polymorphic addition and concatenation.
+
+Adds `value` to `element`.
+The behavior depends on the type of `element`:
+
+- if a number or a string, equivalent to using `+`;
+- if a collection, equivalent to invoking `add` and returning the collection;
+- if an `Appendable` or a `StringBuilder`, equivalent to invoking `append`.
+
+This function is *swapped* and curried. For instance:
+
+    add("b", "a") == "ab"
+
+    list[1, 2, 3]: map(add(3)) == list[4, 5, 6]
+
+    add(42): andThen(add(1337)): andThen(add(69))(list[]) == list[42, 1337, 69]
+
+    add("a", StringBuilder()): toString() == "a"
+
+- *param* `value`: the value to add
+- *param* `element`: the element to add the value to
+----
+@curry
```

Comment:
According `curry` is pure and the only parameter is the decorated function ( i.e. `add` ) is constant, I think you can bang this decoration :wink: